From 153cb753cc6492903f52a0a9a3bd852cdf864a25 Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 12:26:12 -0400
Subject: [ethernet] tg3: fix return value in tg3_get_stats64

Message-id: <1410179173-14304-19-git-send-email-ivecera@redhat.com>
Patchwork-id: 91661
O-Subject: [RHEL7.1 PATCH 18/19] tg3: fix return value in tg3_get_stats64
Bugzilla: 1088637
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: #1088637

Upstream commit(s):
commit 7b31b4deda7679120379c005df8d3d647161eff7
Author: Govindarajulu Varadarajan <_govind@gmx.com>
Date:   Wed Aug 13 13:04:56 2014 +0530

    tg3: fix return value in tg3_get_stats64

    When tp->hw_stats is 0, tg3_get_stats64 should display previously
    recorded stats. So it returns &tp->net_stats_prev. But the caller,
    dev_get_stats, ignores the return value.

    Fix this by assigning tp->net_stats_prev to stats and returning stats.

    Signed-off-by: Govindarajulu Varadarajan <_govind@gmx.com>
    Acked-by: Prashant Sreedharan <prashant@broadcom.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/broadcom/tg3.c b/drivers/net/ethernet/broadcom/tg3.c
index 355cf93..d140f64 100644
--- a/drivers/net/ethernet/broadcom/tg3.c
+++ b/drivers/net/ethernet/broadcom/tg3.c
@@ -14105,8 +14105,9 @@ static struct rtnl_link_stats64 *tg3_get_stats64(struct net_device *dev,
 
  spin_lock_bh(&tp->lock);
  if (!tp->hw_stats) {
+  *stats = tp->net_stats_prev;
   spin_unlock_bh(&tp->lock);
-  return &tp->net_stats_prev;
+  return stats;
  }
 
  tg3_get_nstats(tp, stats);
-- 
1.7.1